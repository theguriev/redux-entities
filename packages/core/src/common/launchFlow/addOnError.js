const converErrorToObject = error => ({
    stack: error.stack,
    message: error.message,
    name: error.name
})

const isError = candidate => candidate instanceof Error || candidate?.constructor?.name.includes('Error')

export default ({
    actions: {
        error
    },
    hash,
    key,
    dispatch,
    setState,
    mountedRef,
    onError
}) => ({
    onError: errorObject => {
        const payload = (
            isError(errorObject) ? converErrorToObject(errorObject) : errorObject
        )
        const action = error({
            payload,
            hash,
            key
        })
        dispatch(action)
        if (mountedRef.current) {
            setState(state => ({
                ...state,
                error: payload,
                isLoading: false,
                isRefetching: false,
                cache: 'cache-first'
            }))
        }
        onError(action)
        return payload
    }
})
