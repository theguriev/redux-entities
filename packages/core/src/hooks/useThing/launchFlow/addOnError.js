import { Types } from '../utils'

export default ({
    toType,
    hash,
    key,
    dispatch,
    setState,
    mountedRef,
    onError
}) => ({
    onError: payload => {
        const action = {
            type: toType(Types.Error),
            payload,
            hash,
            key
        }
        dispatch(action)
        if (mountedRef.current) {
            setState(state => ({
                ...state,
                error: payload,
                isLoading: false,
                isRefetching: false,
                cache: 'cache-first'
            }))
        }
        onError(action)
        return payload
    }
})
