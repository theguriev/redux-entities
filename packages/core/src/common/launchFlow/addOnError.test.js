import {
    toType,
    dispatch,
    setState,
    noop
} from '@redux-things/mocks'
import addOnError from './addOnError'

describe('addOnError', () => {
    test('add onError option | mounted', async () => {
        dispatch.clear()
        setState({ initial: true })
        let onErrorHasBeenCalled = false
        const { onError } = addOnError({
            toType,
            hash: 'hash',
            key: 'TKey',
            dispatch,
            setState,
            mountedRef: { current: true },
            onError: error => {
                onErrorHasBeenCalled = error
            }
        })
        const res = onError({ testing: true })
        expect(res.testing).toBe(true)
        expect(dispatch.actions).toMatchSnapshot()
        expect(setState.state).toMatchSnapshot()
        expect(onErrorHasBeenCalled).toMatchSnapshot()
    })

    test('add onError option | unmounted', async () => {
        setState({ nothing: { should: { change: true } } })
        const { onError } = addOnError({
            toType,
            hash: 'hash',
            key: 'TKey',
            dispatch,
            setState,
            mountedRef: { current: false },
            onError: noop
        })
        onError({ testing: true })
        expect(setState.state).toEqual({ nothing: { should: { change: true } } })
    })
})
