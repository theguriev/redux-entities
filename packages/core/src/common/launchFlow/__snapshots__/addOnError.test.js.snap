// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addOnError add onError option | mounted 1`] = `
Array [
  Object {
    "hash": "hash",
    "key": "TKey",
    "payload": Object {
      "testing": true,
    },
    "type": "error",
  },
]
`;

exports[`addOnError add onError option | mounted 2`] = `
Object {
  "cache": "cache-first",
  "error": Object {
    "testing": true,
  },
  "initial": true,
  "isLoading": false,
  "isRefetching": false,
}
`;

exports[`addOnError add onError option | mounted 3`] = `
Object {
  "hash": "hash",
  "key": "TKey",
  "payload": Object {
    "testing": true,
  },
  "type": "error",
}
`;
